### PR TITLE
IPsec support

### PR DESCRIPTION
**Pull request by ondrej-fabry**
<details>
  <summary>Test data</summary>
  <b>SA</b>
  <pre>
vpp-agent-ctl -put /vnf-agent/agent1/vpp/config/v1/ipsec/sa/sa10 - << EOF
{
	"name":"sa10",
	"spi": 1001,
	"protocol": 1,
	"crypto_alg": 1,
	"crypto_key": "4a506a794f574265564551694d653768",
	"integ_alg": 2,
	"integ_key": "4339314b55523947594d6d3547666b45764e6a58"
}
EOF
vpp-agent-ctl -put /vnf-agent/agent1/vpp/config/v1/ipsec/sa/sa20 - << EOF
{
	"name":"sa20",
	"spi": 1000,
	"protocol": 1,
	"crypto_alg": 1,
	"crypto_key": "4a506a794f574265564551694d653768",
	"integ_alg": 2,
	"integ_key": "4339314b55523947594d6d3547666b45764e6a58"
}
EOF
</pre>
  <b>SPD</b>
<pre>
vpp-agent-ctl -put /vnf-agent/agent1/vpp/config/v1/ipsec/spd/spd1 - << EOF
{
	"name":"spd1",
	"interfaces": [
		{ "name": "GigabitEthernet0/8/0" }
	],
	"policy_entries": [
		{
			"priority": 100,
			"is_outbound": false,
			"action": 0,
			"protocol": 50
		},
		{
			"priority": 100,
			"is_outbound": true,
			"action": 0,
			"protocol": 50
		},
		{
			"priority": 10,
			"is_outbound": false,
			"remote_addr_start": "192.168.100.2",
			"remote_addr_stop": "192.168.100.2",
			"local_addr_start": "192.168.100.3",
			"local_addr_stop": "192.168.100.3",
			"action": 3,
			"sa": "sa20"
		},
		{
			"priority": 10,
			"is_outbound": true,
			"remote_addr_start": "192.168.100.2",
			"remote_addr_stop": "192.168.100.2",
			"local_addr_start": "192.168.100.3",
			"local_addr_stop": "192.168.100.3",
			"action": 3,
			"sa": "sa10"
		}
	]
}
EOF
</pre>
</details>

### CODE REVIEW

File: plugins/defaultplugins/ipsecplugin/ipsec_config.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+		}
+	}
+
+	plugin.SpdIndexes.UnregisterName(oldSpd.Name)
+	plugin.Log.Infof("Deleted SPD %v", oldSpd.Name)
+
+	return nil
+}
+
+// ConfigureSA
```
Comment (VladoLavor) [inline plugins/defaultplugins/ipsecplugin/ipsec_config.go]: add a few words

File: plugins/defaultplugins/ipsecplugin/vppcalls/ipsec_vppcalls.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+		return err
+	}
+	if reply.Retval != 0 {
+		return fmt.Errorf("%s returned %d", reply.GetMessageName(), reply.Retval)
+	}
+
+	return nil
+}
+
+func AddSPD(spdID uint32, vppChan *govppapi.Channel, stopwatch *measure.Stopwatch) error {
```
Comment (VladoLavor) [inline plugins/defaultplugins/ipsecplugin/vppcalls/ipsec_vppcalls.go]: comments? I would expect golint to be unhappy about it

File: plugins/defaultplugins/ipsecplugin/ipsec_config.go [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+// ConfigureSPD configures SPD
+func (plugin *IPSecConfigurator) ConfigureSPD(spd *ipsec.SecurityPolicyDatabases_SPD) error {
+	plugin.Log.Debugf("Configuring SPD %v", spd.Name)
+
+	spdID := plugin.SpdIndexSeq
+	plugin.SpdIndexSeq++
+
+	for _, entry := range spd.PolicyEntries {
+		if entry.Sa != "" {
```
[the remark this answers is not in the document]
Comment (ondrej-fabry) [reply plugins/defaultplugins/ipsecplugin/ipsec_config.go]: Not mandatory.

File: plugins/defaultplugins/ipsecplugin/ipsec_config.go [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+	return nil
+}
+
+// DeleteSPD
+func (plugin *IPSecConfigurator) DeleteSPD(oldSpd *ipsec.SecurityPolicyDatabases_SPD) error {
+	plugin.Log.Debugf("Deleting SPD %v", oldSpd.Name)
+
+	spdID, _, exists := plugin.SpdIndexes.LookupIdx(oldSpd.Name)
+	if !exists {
+		plugin.Log.Warnf("SPD %q not found", oldSpd.Name)
```
[the remark this answers is not in the document]
Comment (ondrej-fabry) [reply plugins/defaultplugins/ipsecplugin/ipsec_config.go]: I think this case is similar to removing non-existing bridge domain: https://github.com/ligato/vpp-agent/blob/pantheon-dev/plugins/defaultplugins/l2plugin/bd_config.go#L216 where it isn't unregistered or cached either.

File: plugins/defaultplugins/ipsecplugin/ipsec_config.go [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+		return nil
+	}
+	if err := vppcalls.DelSPD(spdID, plugin.vppCh, plugin.Stopwatch); err != nil {
+		return err
+	}
+
+	// remove cache entries related to the SPD
+	for i, entry := range plugin.SPDIfCache {
```
[the remark this answers is not in the document]
Comment (ondrej-fabry) [reply plugins/defaultplugins/ipsecplugin/ipsec_config.go]: You might be confusing two types of caches. There is cache for SPDs and cache for assignments of SPD to interface. SPD can exist without any association, that's why SPD to interface associations are cached separately when interface doesn't exists or is removed. The other cache for SPDs is used when it's SA dependency doesn't exist yet or is removed.

However I still forgot to clear SPD from cache on DeleteSPD, will add that.

[the remark this answers is not in the document]
Comment (VladoLavor) [reply plugins/defaultplugins/ipsecplugin/ipsec_config.go]: ok, got it

**Pull request merged**